crates/primitives/src/proofs.rs

```diff
@@ -70,18 +70,55 @@ pub fn calculate_withdrawals_root(withdrawals: &[Withdrawal]) -> B256 {
 
 /// Calculates the receipt root for a header.
 pub fn calculate_receipt_root(receipts: &[ReceiptWithBloom]) -> B256 {
+    #[cfg(feature = "optimism")]
+    {
+        // There is a minor bug in op-geth and op-erigon where in the Regolith hardfork,
+        // the receipt root calculation does not include the deposit nonce in the receipt
+        // encoding. This will be fixd in the next hardfork, however for now, we must strip
+        // the deposit nonce from the receipts before calculating the receipt root.
+        let receipts = receipts
+            .iter()
+            .cloned()
+            .map(|mut r| {
+                r.receipt.deposit_nonce = None;
+                r
+            })
+            .collect::<Vec<_>>();
+
+        ordered_trie_root_with_encoder(receipts.as_slice(), |r, buf| r.encode_inner(buf, false))
+    }
+
+    #[cfg(not(feature = "optimism"))]
     ordered_trie_root_with_encoder(receipts, |r, buf| r.encode_inner(buf, false))
 }
 
-/// Calculates the receipt root for a header for the reference type of [ReceiptWithBloom].
+/// Calculates the receipt root for a header for the reference type of [Receipt].
 ///
 /// NOTE: Prefer [calculate_receipt_root] if you have log blooms memoized.
-pub fn calculate_receipt_root_ref<T>(receipts: &[&T]) -> B256
-where
-    for<'a> ReceiptWithBloomRef<'a>: From<&'a T>,
-{
+pub fn calculate_receipt_root_ref(receipts: &[&Receipt]) -> B256 {
+    #[cfg(feature = "optimism")]
+    {
+        // There is a minor bug in op-geth and op-erigon where in the Regolith hardfork,
+        // the receipt root calculation does not include the deposit nonce in the receipt
+        // encoding. This will be fixd in the next hardfork, however for now, we must strip
+        // the deposit nonce from the receipts before calculating the receipt root.
+        let receipts = receipts
+            .iter()
+            .map(|r| {
+                let mut r = (*r).clone();
```

Comment:
this can be expensive :/ i dont like it but if you're saying it'll be fixed in 2 weeks w/ Canyon maybe fine

Reply:
Yeah this is very unideal from a performance standpoint, but unfortunately the bug is canonical. This issue is fixed in Canyon, see #5210

Reply:
We can modify this route to take temporary ownership of the receipts, strip deposit nonces, compute the root, and then re-hydrate them and pass them back. I figured that this would be more invasive given that the non-optimism path doesn't need this, will defer to you guys.

Reply:
Think @clabby is right here. It's probably not worth touching more reth-bin logic just to optimize the op-reth codepath for now. We can improve this later.

Reply:
@clabby @refcell Might be out of topic but when would span-batch happen using Ethereum as a DA layer getting expensive time by time

Reply:
So, let's keep discussion here op-reth pr specific, but would be happy to discuss in an issue in the monorepo. Could you tag us there so we don't lose track?